packages/scanner/src/rules/jwtUnverifiedSignature.ts

```diff
@@ -0,0 +1,49 @@
+import { MatchResult, Rule, RuleLogic } from '../types';
+import parseRuleDescription from './lib/parseRuleDescription';
+import { Event } from '@appland/models';
+
+class JwtUnverifiedSignatureLogic implements RuleLogic {
+  matcher(event: Event): MatchResult[] | undefined {
+    let verified = false;
+    const matches = new Array<MatchResult>();
+
+    const children = [...event.descendants()];
+    while (children.length) {
+      const child = children.pop() as Event;
+      if (child.labels.has('jwt.signature.verify')) {
```

Comment:
Is this sure to occur only in a descendent or could it also occur in a preceding event on the same receiver object?

Reply:
I've updated the conditions to identify a signature verification match. An event which is labelled `jwt.signature.verify` must either:
- Also contain the `jwt.decode` label
```ts
function decode(token: string): Jwt | undefined {
  const [header, payload, signature] = token.split('.').map(t => btoa(t));

  switch (header.alg.toLowerCase()) {
  case 'hs256'
    // The token signature is validated here inline during decoding.
    break;
  
  case 'none':
  default:
    return undefined;
  }
  
  return new Jwt(header, payload, signature);
}
```
- Have the same receiver (by looking at object ID)
```ts
class Jwt {
  constructor(protected readonly token: string) {
    this.decode();
    this.verifySignature();
  }

  decode() {
    // Parse token segments
  }
  
  verifySignature() {
    // Make sure the signature matches
  }
}
```
- Contain a parameter which matches the decoded token string
```ts
class Jwt {
  constructor(protected readonly token: string, key: string) {
    this.decode(token);
    Signature.verify(this.algorithm, key, this.signingInput, this.signature);
  }
}

class Signature {
  static verify(algorithm: Algorithm, key: string, signingInput: string, signature: string) {
    // Perform the signature verification
  }
}
```
- Contain a parameter which matches the receiver of the decode event
```ts
class Jwt implements Signed {
  constructor(protected readonly token: string) {
    this.decode(token);
    Signature.verify(this);
  }
  
  public get signingInput(): string { return `${this.header}.${this.payload}`; }
  public get algorithm(): Algorithm { return Algorithms[this.decodedHeader.alg]; }
}

class Signature {
  static verify(token: Signed, key: string) {
    // Perform the signature verification
  }
}
```

Reply:
It won't look at preceding events, only events which follow the decode. The reason being the token header must first be decoded to retrieve the encryption algorithm, so it's unlikely that the validation can come first.